File: 360.java
/**

360. Sliding Window Median
https://www.lintcode.com/problem/sliding-window-median/description

**/
public class Solution {
    /**
     * @param nums: A list of integers
     * @param k: An integer
     * @return: The median of the element inside the window at each moving
     */
    public List<Integer> medianSlidingWindow(int[] nums, int k) {
        // write your code here

        if (nums == null || nums.length == 0) {
            return new ArrayList<Integer>();
        }

        int n = nums.length;
        PriorityQueue<Integer> max_queue = new PriorityQueue(n, Collections.reverseOrder());
        PriorityQueue<Integer> min_queue = new PriorityQueue(n);

        List<Integer> results = new ArrayList<Integer>();

        int right = 0;
        for (int left = 0; left < n; ++left) {
            while (right < n && right - left < k) {
                if (max_queue.size() == 0 || nums[right] <= max_queue.peek()) {
                    max_queue.offer(nums[right]);
                } else {
                    min_queue.offer(nums[right]);
                }
                balance(max_queue, min_queue);
                ++right;
            }
            if (right - left == k) {
                results.add(max_queue.peek());
            }
            if (right >= n) {
                break;  //because even if right == n, the result is already added in previous statement
            }

            if (nums[left] <= max_queue.peek()){
                max_queue.remove(nums[left]);
            } else {
                min_queue.remove(nums[left]);
            }
            balance(max_queue, min_queue);
        }
        return results;
    }

    private void balance(PriorityQueue<Integer> max_queue, PriorityQueue<Integer> min_queue) {
        while (max_queue.size() < min_queue.size()) {
            max_queue.offer(min_queue.poll());
        }
        while (max_queue.size() - min_queue.size() > 1) {
            min_queue.offer(max_queue.poll());
        }
    }
}

File: lintcode/360.py
"""
360. Sliding Window Median
https://www.lintcode.com/problem/sliding-window-median/description
J's method: https://www.jiuzhang.com/solution/sliding-window-median/#tag-other-lang-python
"""
import heapq
class Heap:
    def __init__(self):
        self.heap = []
        self.deleted = {}
        self.length = 0

    def push(self, element):
        heapq.heappush(self.heap, element)
        self.length += 1

    def pop(self):
        self.remove_redundant()
        self.length -= 1
        return heapq.heappop(self.heap)

    def remove(self, element):
        self.deleted[element] = self.deleted.get(element, 0) + 1
        self.length -= 1

    def remove_redundant(self):
        while self.heap and self.deleted.get(self.heap[0]):
            removed_duplicate = heapq.heappop(self.heap)
            self.deleted[removed_duplicate] -= 1
            if self.deleted[removed_duplicate] == 0:
                del self.deleted[removed_duplicate]

    def peek(self):
        self.remove_redundant()
        return self.heap[0]

    def __len__(self):
        return self.length

    def __repr__(self):
        return repr(self.heap)

class Solution:
    """
    @param nums: A list of integers
    @param k: An integer
    @return: The median of the element inside the window at each moving
    """
    def medianSlidingWindow(self, nums, k):
        # write your code here
        max_heap = Heap()
        min_heap = Heap()
        removed_elements = set()

        right = 0
        results = []
        n = len(nums)
        for left in range(n):
            while right < n and right - left < k:
                if len(max_heap) == 0 or nums[right] <= -max_heap.peek():
                    max_heap.push(-nums[right])
                else:
                    min_heap.push(nums[right])
                self.balance(max_heap, min_heap)
                right += 1

            if right - left == k:
                results.append(-max_heap.peek())
            if right >= n:
                break;

            if nums[left] <= -max_heap.peek():
                max_heap.remove(-nums[left])
            else:
                min_heap.remove(nums[left])
            self.balance(max_heap, min_heap)

        return results

    def balance(self, max_heap, min_heap):
        while len(max_heap) < len(min_heap):
            max_heap.push(-min_heap.pop())

        while len(max_heap) > len(min_heap) + 1:
            min_heap.push(-max_heap.pop())
